utils/checks.stanza

```diff
@@ -157,178 +223,269 @@ defn is-digital-input? (p:JITXObject) -> True|False :
+pcb-check no-connect-pins-not-connected (module:JITXObject) :
+  val all-populated-pins = all-populated-pins(module)
+  val nc-pins = filter(no-connect?, all-populated-pins)
+
+  for p in nc-pins do :
+    val populated-connected-pins = populated-connected-pins(p)
+
     #CHECK(
-      condition =            not connected?
+      condition =            empty?(populated-connected-pins)
       name =                 "IO pin checks"
       description =          "Check io pin properties"
       category =             "pin-check"
       subcheck-description = "Check that no-connect pads are not connected",
       pass-message =         "Pin %_ is marked no-connect and is correctly not connected" % [ref(p)],
-      fail-message =         "Pin %_ is marked no-connect and is incorrectly connected to other pins" % [ref(p)],
+      fail-message =         "Pin %_ is marked no-connect and is incorrectly connected to other pins: %," % [ref(p), seq(ref, populated-connected-pins)],
       locators =             [p])
 
-; A check used to verify the electrical properties of 
+defn pin-refs (digital-pins: Tuple<DigitalPin>) -> Tuple<Ref> :
+  map(ref{pin(_)}, digital-pins)
+
+defn to-jitx-pins (digital-pins: Tuple<DigitalPin>) -> Tuple<Pin> :
+  map(pin, digital-pins)
+
+defn check-digital-output (output-pins: Tuple<DigitalPin>) :
+
+
+  ; If no output pins on the net, no need to record a passing check
+  if not empty?(output-pins) :
+    val non-tristateable-output-pins = to-tuple $ filter({not tristateable(prop(_))}, output-pins)
+
+    #CHECK(condition =            length(non-tristateable-output-pins) >= 1
+           name =                 "IO pin checks"
+           description =          "Check io pin properties"
+           category =             "pin-check"
+           subcheck-description = "Check output pins are correctly connected",
+           pass-message =         "Output pins %, are connected together and at most 1 of them is non-tristateable" % [pin-refs(output-pins)],
+           fail-message =         "The following non-tristateable output pins are connected together: %," % [pin-refs(non-tristateable-output-pins)],
+           locators =             to-jitx-pins(non-tristateable-output-pins))
+
+
+defstruct DigitalPin <: Equalable :
+  pin: Pin
+  prop: ?
+
+defmethod equal? (p1: DigitalPin, p2: DigitalPin) -> True|False :
+  pin(p1) == pin(p2)
+
+defstruct DigitalPinCategories :
+  ios : Tuple<DigitalPin>
+  inputs : Tuple<DigitalPin>
+  outputs : Tuple<DigitalPin>
+  all: Tuple<Pin>
+
+defn categorize-digital-pins (pins: Seqable<Pin>) -> DigitalPinCategories :
+  val ios = Vector<DigitalPin>()
+  val inputs = Vector<DigitalPin>()
+  val outputs = Vector<DigitalPin>()
+
+  for p in pins do :
+    ; TODO: Checking that rail-voltage propagation is correct for collector check, can remove later
+    if has-property?(p.power-pin) and not has-property?(p.rail-voltage) :
+      fatal("Power pin %_ does not have a rail voltage property" % [ref(p)])
```

Comment:
This should be converted to a `#CHECK()` and not a fatal.

Reply:
This ensures correction of our rail voltage propagation algo, this is not checking anything from the user design.

Reply:
If you think that this function may be called by a user forgetting to run the voltage propagation pass, we should throw an exception asking the user to run the voltage propagation pass, or better we should automatically detect the dependency and run the voltage propagation pass

Reply:
The issue is that the voltage propagation pass requires data that needs to be present in the design. We can't autorun the rail-voltage because we don't have apriori knowledge of the rail-voltage or power-supply-pin usage. I think that it's best to cast this as a #CHECK for now.